utility/publish.go

```diff
@@ -0,0 +1,316 @@
+		log.Infof("Update listing finished")
+	}
+	return nil
+}
+
+// readLocalisedRecentChanges reads the recent changes from the given path and returns them as a map.
+func readLocalisedRecentChanges(recentChangesDir string) (map[string]string, error) {
+	recentChangesMap := map[string]string{}
+
+	pattern := filepath.Join(recentChangesDir, "whatsnew-*-*")
+	recentChangesPaths, err := filepath.Glob(pattern)
+	if err != nil {
+		return map[string]string{}, err
+	}
+
+	pattern = `whatsnew-(?P<local>.*-.*)`
+	re := regexp.MustCompile(pattern)
+
+	for _, recentChangesPath := range recentChangesPaths {
+		matches := re.FindStringSubmatch(recentChangesPath)
+		if len(matches) == 2 {
+			language := matches[1]
+			content, err := fileutil.ReadStringFromFile(recentChangesPath)
+			if err != nil {
+				return map[string]string{}, err
+			}
+
+			recentChangesMap[language] = content
+		}
+	}
+	if len(recentChangesMap) > 0 {
+		log.Debugf("Found the following recent changes:")
+		for language, recentChanges := range recentChangesMap {
+			log.Debugf("%v\n", language)
+			log.Debugf("Content: %v", recentChanges)
+		}
+	} else {
+		log.Debugf("No recent changes found")
+	}
+
+	return recentChangesMap, nil
+}
+
+// GetRelease gets a release from a track based on it's status.
+func GetRelease(config config.Configs, releases *[]*androidpublisher.TrackRelease) *androidpublisher.TrackRelease {
+	status := GetReleaseStatusFromConfig(config)
+	for _, release := range *releases {
+		if status == release.Status {
+			return release
+		}
+	}
+	newRelease := androidpublisher.TrackRelease{
+		Status: status,
+	}
+	if config.UserFraction != 0 {
+		newRelease.UserFraction = config.UserFraction
+	}
+	*releases = append(*releases, &newRelease)
+	return &newRelease
+}
+
+// UpdateRelease creates and returns a new release object with the given version codes.
+func UpdateRelease(configs config.Configs, versionCodes googleapi.Int64s, releases *[]*androidpublisher.TrackRelease) error {
+	release := GetRelease(configs, releases)
+	if configs.UntrackBlockingVersions {
+		removeBlockingVersionFromRelease(release, versionCodes)
+	} else {
+		release.VersionCodes = append(release.VersionCodes, versionCodes...)
+	}
+	log.Infof("Release version codes are: %v", release.VersionCodes)
+
+	log.Printf(" assigned app versions: %v", release.VersionCodes)
+	if err := updateListing(configs, release); err != nil {
+		return fmt.Errorf("failed to update listing, reason: %v", err)
+	}
+	return nil
+}
+
+// removeBlockingVersionFromRelease removes blocking version from a given version, which would shadow the given version.
+func removeBlockingVersionFromRelease(release *androidpublisher.TrackRelease, newVersionCodes googleapi.Int64s) {
+	log.Printf("Checking app versions on release: %s", release.Name)
+	log.Infof("Current version codes: %v", release.VersionCodes)
+	log.Infof("New version codes: '%v'", newVersionCodes)
+
+	var cleanTrack bool
+	if len(release.VersionCodes) != len(newVersionCodes) {
```

Comment:
Why the number of versioncodes matters?

Reply:
This way if the number of uploaded apps changed, only the newly uploaded apps will be on the store. This is a legacy code/function.